From c4aacd3965621e6c5c8f0b837a9e0fb9e9523be2 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:13:42 -0400
Subject: [powerpc] powernv: Check for IRQHAPPENED before sleeping

Message-id: <1410545655-205645-594-git-send-email-dzickus@redhat.com>
Patchwork-id: 94526
O-Subject: [RHEL7 PATCH 593/626] powerpc/powernv: Check for IRQHAPPENED before sleeping
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit c733cf83bbd2b0978456822cdac8c726085dcc14
Author: Preeti U Murthy <preeti@linux.vnet.ibm.com>
Date:   Wed Jul 2 09:19:35 2014 +0530

    powerpc/powernv: Check for IRQHAPPENED before sleeping

    Commit 8d6f7c5a: "powerpc/powernv: Make it possible to skip the IRQHAPPENED
    check in power7_nap()" added code that prevents cpus from checking for
    pending interrupts just before entering sleep state, which is wrong. These
    interrupts are delivered during the soft irq disabled state of the cpu.

    A cpu cannot enter any idle state with pending interrupts because they will
    never be serviced until the next time the cpu is woken up by some other
    interrupt. Its only then that the pending interrupts are replayed. This can result
    in device timeouts or warnings about this cpu being stuck.

    This patch fixes ths issue by ensuring that cpus check for pending interrupts
    just before entering any idle state as long as they are not in the path of split
    core operations.

    Signed-off-by: Preeti U Murthy <preeti@linux.vnet.ibm.com>
    Acked-by: Michael Ellerman <mpe@ellerman.id.au>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kernel/idle_power7.S b/arch/powerpc/kernel/idle_power7.S
index 2480256..5cf3d36 100644
--- a/arch/powerpc/kernel/idle_power7.S
+++ b/arch/powerpc/kernel/idle_power7.S
@@ -131,7 +131,7 @@ _GLOBAL(power7_nap)
 
 _GLOBAL(power7_sleep)
  li r3,1
- li r4,0
+ li r4,1
  b power7_powersave_common
  /* No return */
 
-- 
1.7.1